### ISSUE TITLE
For certain packages, `pixi add` leaves failed dependency lines in pixi.toml

### ISSUE THREAD

**Issue by etjones-erisyon** [prefix-dev/pixi#526]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://github.com/prefix-dev/pixi/releases) of pixi, using `pixi --version`.

### Reproducible example

Running pixi v0.9.1 on my Arm64 Mac M1, MacOS 13.3:

```shell
https://github.com/ruben-arts/turtlesim-pixi.git
cd turtlesim-pixi
pixi add ros-humble-rosbridge-server

```
I see several downloads, then the following error: 
```shell
  × Cannot solve the request because of: No candidates were found for ros-humble-rosbridge-server 1.2.0.*.
  │ 
```

That's fine; packages fail to install sometimes. But `pixi.toml` now contains:
```toml
ros-humble-rosbridge-server = "1.2.0.*"
```

And any future invocations of pixi (`run`, `add`, `shell` etc) return the same `Cannot solve the request` error until I manually removed the dependency line.

### Issue description

I think the example above covers it.

### Expected behavior

Optimal behavior would be if the package (in this case `ros-humble-rosbridge-server`) installed correctly.
Next best would be if whatever process yielded the `  × Cannot solve the request ...` error also removed the dependency line from `pixi.toml`.  `pixi.toml` should be unchanged except in case of a successfully resolved install.

**Comment by baszalmstra** [comment]
Yes thats definitely incorrect! Thanks for reporting!

**Comment by baszalmstra** [comment]
The issue here is twofold. First and foremost is that the project file should not be written if `add` leaves the project in a broken/non-solvable state. I will create a PR that saves the project after creating the lock-file to fix this.

The second issue is that the error that occurs is unexpected. Add first tries to solve the environment for every platform individually, then checks the version of the requested package and then adds a strict dependency on the lowest encountered version of that package. The reasoning here is that this will make sure that the lowest version will be solvable for every platform. Unfortunately for some packages like [`ros-humble-rosbridge-server`](https://prefix.dev/channels/robostack-humble/packages/ros-humble-rosbridge-server) version 1.2.0 is only available for linux and version 1.3.0 is only available for macOS. The code then selects 1.2.0 as the lowest version between all platforms and pins that. The problem is that that makes macOS unsolvable because 1.2.0 is not available on that platform. I will create a PR that adds a looser bound if that happens. (e.g. `>=1.2.0,<1.4.0`). Im going to combine this behavior with implementing the behavior described in: https://github.com/prefix-dev/pixi/issues/285.

**Comment by etjones-erisyon** [comment]
Really great to hear what the problem was and how it was solved. Thanks guys!
